[stm32] unify f0/f3 i2c

geomatsi opened the pull request:
Add i2c support for f0 series. Both f0 and f3 series share the same i2c hardware,
so share its driver among them.

Tested on a custom f0 board. Both libopencm3 and examples compile-tested for f3.

Signed-off-by: Sergey Matyukevich <geomatsi@gmail.com>

karlp commented:
This sounds all good, I've not looked at in much detail, but isn't this the also used on l0/l4?  (and perhaps f7 too?)  I'd rather see this named i2c-v2 or similar, like I did with the adc recently, rather than this soup of f03 names.

geomatsi commented:
Well, v1/v2 approach looks more sane. Though I am not sure about i2c-v2 on other families. So far I could find [this ST document](http://www.st.com/content/ccc/resource/technical/document/user_manual/0e/49/a3/22/24/7d/48/6c/DM00063145.pdf/files/DM00063145.pdf/jcr:content/translations/en.DM00063145.pdf) which mentions f0 and f3 families. Maybe it worth looking at CPAL code for the answers.

zyp commented:
Just check the reference manuals for l0, l4 and f7 and see whether the register layout is matching.

geomatsi commented:
Make sense. Ok, I will split current i2c headers and libs into v1 and v2, and then re-submit. Closing down this PR.

karlp commented:
cool.  fwiw, you can also keep the PR, and simply push -f a new branch in it's place, it keeps the discussion.  Sometimes that's helpful, sometimes just noise of course :)

geomatsi commented:
Ok, if you don't object against forced PR updates, then I will keep this one. Thanks.

geomatsi commented:
Here is an updated PR. It is now split into two commits. The first commit unifies i2c implementations using available i2c_common_all code for "v1" and i2c implementation for stm32/f3  for "v2". The second commit adds "v2" i2c support for l0 and l4 chip families: according to registry maps they share i2c implementation with f0 and f3.

Actual testing has been done on a custom f0 board. Both libopencm3 and examples have been compile-tested.

karlp commented on lib/stm32/common/i2c_common_v2.c:
```diff
@@ -0,0 +1,477 @@
+	case I2C2:
+		rcc_peripheral_reset(&RCC_APB1RSTR, RCC_APB1RSTR_I2C2RST);
+		rcc_peripheral_clear_reset(&RCC_APB1RSTR, RCC_APB1RSTR_I2C2RST);
+		break;
```
Potentially
```
rcc_periph_reset_pulse(RCC_I2C1)
```
Is simpler and shorter, but they're equivalent really.

geomatsi replied:
Nice. I will add to another patch to the series with i2c-v2 cleanup.

geomatsi replied:
Besides, I2C3 and I2C4 are missed in this reset function. To be added:
```
#if defined(I2C3_BASE)
        case I2C3:
                rcc_periph_reset_pulse(RCC_I2C3);
                break;
#endif
#if defined(I2C4_BASE)
        case I2C4:
                rcc_periph_reset_pulse(RCC_I2C4);
                break;
#endif
```

karlp replied:
sounds good

geomatsi commented on include/libopencm3/stm32/l4/i2c.h:
```diff
@@ -0,0 +1,48 @@
+/* for API compatibility */
+#define RCC_APB1RSTR			RCC_APB1RSTR1
+#define RCC_APB1RSTR_I2C2RST		RCC_APB1RSTR1_I2C2RST
+#define RCC_APB1RSTR_I2C1RST		RCC_APB1RSTR1_I2C2RST
```
It was for i2c_reset only. It is not needed when rcc_periph_reset_pulse is used. Will be removed.

geomatsi commented:
PR has been updated.

v2 -> v3 changes:
- add cleanup patch:
  - modify i2c-v2 reset functon: use rcc_periph_reset_pulse
  - use ifdef's to drop unavailable i2c units from i2c_reset in i2c-v2
  - add I2C4 to i2c-v1
- get rid of API compatibility hack in L4 i2c header

ChuckM commented on lib/stm32/common/i2c_common_v1.c:
```diff
@@ -65,6 +65,11 @@ void i2c_reset(uint32_t i2c)
+#if defined(I2C4_BASE)
+	case I2C4:
+		rcc_periph_reset_pulse(RCC_I2C4);
+		break;
```
This reminds me once again of the enum problem. @karlp suggested you add `rcc_periph_reset_pulse(RCC_xxx)` but the actual enum constant **should be** RST_xxx which is shown above in line 65. This confusion between RCC_periph and RST_periph is a ripe source of errors in the code base.

ChuckM commented on lib/stm32/common/i2c_common_v2.c:
```diff
@@ -38,12 +38,24 @@ void i2c_reset(uint32_t i2c)
 	case I2C1:
-		rcc_peripheral_reset(&RCC_APB1RSTR, RCC_APB1RSTR_I2C1RST);
-		rcc_peripheral_clear_reset(&RCC_APB1RSTR, RCC_APB1RSTR_I2C1RST);
+		rcc_periph_reset_pulse(RCC_I2C1);
```
Change to RST_I2C1

ChuckM commented on lib/stm32/common/i2c_common_v2.c:
```diff
@@ -38,12 +38,24 @@ void i2c_reset(uint32_t i2c)
+		rcc_periph_reset_pulse(RCC_I2C4);
+		break;
+#endif
+	default:
```
Replace all RCC_xxx instances with RST_xxx instances

geomatsi commented:
v3 -> v4

Fix typo in i2c_reset functions: RCC_I2CX -> RST_I2CX.

geomatsi commented:
Any further comments/suggestions ?

geomatsi commented:
v4 -> v5: no changes, rebased on top of master.

Hi @karlp and all,
It looks like I addressed all the comments so far. There was an additional request from Chuck to unify RCC functions to take one enum. Sounds good but I think it should be implemented and merged as a separate PR. Is there anything else I can do for this particular set of i2c changes in order to get it merged ? 

Thanks,
Sergey

allesblinkt commented:
I hope this can be incorporated into libopencm3/main soon. I have been using this branch for development of a project already for quite some time without any issues. Thank you. #

geomatsi commented:
v5 -> v5: no actual updates, rebase on top of master and make sure that everything is ok

karlp commented:
Started my i2c tests tonight (finally)

karlp commented:
merged into https://github.com/karlp/libopencm3/tree/karlp-staging-2017.1  just working on some f0/l0/f3 i2c master code to go along with the f4/l1 code I have already for coverage & regression testing

karlp commented:
I've removed the write_i2c and read_i2c routines as they were a bit limiting and added similar but extended functions in my branch that has this merged for staging: https://github.com/libopencm3/libopencm3/pull/737

comments welcome.
